README.md
<!-- [![doc][1]][2] [![matrix][12]][13] -->

![GitHub release][3] [![build][6]][7] [![license][4]][5] [![grade][8]][9] [![coverage][10]][11] [![discord][14]][15]

# Pendulum

Pendulum is a quorum based [Tangle](https://github.com/iotaledger/iri/) implementation designed towards reliable timekeeping and high-throughput messaging.

-   **Latest release:** 1.0.7 release
-   **License:** GPLv3

Special thanks to all of the [IOTA Contributors](https://github.com/iotaledger/iri/graphs/contributors)!

## Hardware requirements

**Minimal** (~t2.small AWS instance)
  - 2GB RAM
  - 1 GHz CPU
  - 10 GB storage
  - 10Mbit/s WAN, static IP

**Optimal** (~t2.medium AWS instance)
  - 4GB RAM or more
  - 2 or more 2GHz CPU cores (~ t2.medium AWS instance)
  - 50GB SSD
  - 1Gbit/s WAN, static IP

**Enterprise-grade**
  - Four or more instances with Optimal specs
    - two or more instances with `--remote` API enabled
    - two or more "relayer" instances connected to multiple peers  
  - For validators: additional dedicated instance with Optimal specification for the validator node
  - HA loadbalancer proxing API instances. Can be hardware or software based (e.g. [Nginx cluster sample config](#nginx-cluster-sample-config) below)


## Developers

-   Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.
-   Please read the [pendulum-1.0-specifications](https://github.com/HelixNetwork/helix-specs/tree/master/specs/1.0) before contributing.

## Installing

Make sure you have [**Maven**](https://maven.apache.org/) and [**Java 8**](https://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html) installed on your computer.

### Download

    $ git clone https://github.com/HelixNetwork/pendulum.git

### Build

Build an executable jar at the `target` directory using maven.

    $ cd pendulum
    $ mvn clean package

### Launch Full node

    java -jar target/pendulum-<VERSION>.jar -p 8085

### Launch Validator node
Launching a node as a validator first requires to generate a 64 character hex string, that is used as a seed for key generation. You will find the public key in the last line of the `validator.key` file contained in the resources directory. If you wish to act as a validator, please send a request to dt@hlx.ai containing your public key.

    java -jar target/pendulum-<VERSION>.jar -p 8085 --validator <pathToValidatorSeed>


### Nginx cluster sample config

For production-level applications we recommend exposing a single public API endpoint reverse-proxing multiple fullnode instances. Additionally, we highly recommend obtaining an SSL certificate from a trusted authority (e.g. from Let’s Encrypt).

Below is a sample configuration file for the popular Nginx webserver (typically put into `/etc/nginx/conf.d/` ). For more information please consult the official Nginx [documentation](https://docs.nginx.com/nginx/admin-guide/web-server/reverse-proxy/)

```
upstream pendulum {
        ip_hash;
        Server fullnode1.ip.address:8085;
        Server fullnode2.ip.address:8085;
}


server {
        listen 443 ssl;
        listen [::]:443 ssl;
        server_name my.api.endpoint.com;

        server_tokens off;

        ssl_certificate /path/to/cert.pem;
        ssl_certificate_key /path/to/key.pem;
        ssl_session_timeout 5m;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_ciphers 'EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH';
        ssl_prefer_server_ciphers on;

        ssl_ecdh_curve secp384r1;
        ssl_session_tickets off;

        # OCSP stapling
        ssl_stapling on;

        ssl_stapling_verify on;
        resolver 8.8.8.8;


        location / {
                ## CORS
                proxy_hide_header Access-Control-Allow-Origin;
                add_header 'Access-Control-Allow-Origin' '*' always;
                add_header 'Access-Control-Allow-Credentials' 'true';
                add_header 'Access-Control-Allow-Headers' 'Authorization,Accept,Origin,DNT,X-HELIX-API-Version,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range';
                add_header 'Access-Control-Allow-Methods' 'GET,POST,OPTIONS,PUT,DELETE,PATCH';

                if ($request_method = 'OPTIONS') {
                        add_header 'Access-Control-Allow-Origin' '*';
                        add_header 'Access-Control-Allow-Credentials' 'true';
                        add_header 'Access-Control-Allow-Headers' 'Authorization,Accept,Origin,DNT,X-HELIX-API-Version,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Content-Range,Range';
                        add_header 'Access-Control-Allow-Methods' 'GET,POST,OPTIONS,PUT,DELETE,PATCH';
                        add_header 'Access-Control-Max-Age' 1728000;
                        add_header 'Content-Type' 'text/plain charset=UTF-8';
                        add_header 'Content-Length' 0;
                        return 204;
                }

                proxy_redirect off;
                proxy_set_header host $host;
                proxy_set_header X-real-ip $remote_addr;
                proxy_set_header X-forward-for $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;
                proxy_pass http://pendulum;
        }
}

```


## Configuration

| Option                   | Short | Description                                                                                                                                 | Example Input                                                       |
| ------------------------ | ----- | ------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------- |
| `--port`                 | `-p`  | This is a _mandatory_ option that defines the port to be used to send API commands to your node                                             | `-p 8085`                                                           |
| `--neighbors`            | `-n`  | Neighbors that you are connected with will be added via this option.                                                                        | `-n "udp://148.148.148.148:4100 udp://[2001:db8:a0b:12f0::1]:4100"` |
| `--config`               | `-c`  | Config INI file that can be used instead of CLI options. See more below                                                                     | `-c x.ini`                                                          |
| `--udp-receiver-port`    | `-u`  | UDP receiver port                                                                                                                           | `-u 4100`                                                           |
| `--tcp-receiver-port`    | `-t`  | TCP receiver port                                                                                                                           | `-t 5100`                                                           |
| `--ms-delay`             | `-m`  | Sets delay for auto-milestones.                                                                                                             | `-m 60`                                                             |
| `--testnet`              |       | Testnet flag, bypasses milestone signature validation and pow difficulty.                                                                   | `--testnet`                                                         |
| `--remote`               |       | Remotely access your node and send API commands                                                                                             | `--remote`                                                          |
| `--remote-auth`          |       | Require authentication password for accessing remotely. Requires a correct `username:hashedpassword` combination passed to the Auth Header. | `--remote-auth token:<your_token>`                             |
| `--remote-limit-api`     |       | Exclude certain API calls from being able to be accessed remotely                                                                           | `--remote-limit-api "attachToTangle, addNeighbors"`                 |
| `--send-limit`           |       | Limit the outbound bandwidth consumption. Limit is set to mbit/s                                                                            | `--send-limit 1.0`                                                  |
| `--max-peers`            |       | Limit the number of max accepted peers. Default is set to 0.                                                                                | `--max-peers 8`                                                     |
| `--dns-resolution-false` |       | Ignores DNS resolution refreshing                                                                                                           | `--dns-resolution-false`                                            |
| `--savelog-enabled`      |       | Writes the log to file system                                                                                                               | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
| `--validator`            |       | Flag that enables applying as a validator in the network. A path to a file containing the seed has to be passed.                                                                                                                | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
| `--update-validator`     |       | The desired delay for updating validators in seconds.                                                                                                            | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
| `--start-validator`      |       | The number of rounds between validators are published and the round they start to operate.                                                                                                                | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
| `--genessis`             |       | Time when the ledger started.                                                                                                                | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
| `--round`                |       | Duration of a round in milli secounds.                                                                                                                | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
| `--round-pause`          |       | Duration of time to finalize the round in milli secounds.                                                                                                                | `--savelog-enabled`                                                 |                                                                      | `--pow-disabled`                                                    |
                                                                                                          | `--dns-resolution-false`     |


### INI

You can also provide an ini file to store all of your command line options and easily update (especially neighbors) if needed. You can enable it via the `--config` flag. Here is an example INI file:

```text
[HLX]
PORT = 8085
UDP_RECEIVER_PORT = 4100
NEIGHBORS = udp://my.favorite.com:5100
HXI_DIR = XI
DB_PATH = db
ZMQ_ENABLED = true
```

## MessageQ

MessageQ is a small zmq wrapper for streaming gathered metrics and statistics of topics, enabling targeted event streams from subscribing clients to processes of the node.
A client interested in real time state updates and notifications could use any desired [zmq-client](https://github.com/zeromq/zeromq.js/) to start listening to topics.

Currently the following topics are covered:

| Topic       | Description                                                             |
| ----------- | ----------------------------------------------------------------------- |
| `dns`       | Neighbor related info                                                   |
| `hmr`       | Hit/miss ration                                                         |
| `antn`      | Added non-tethered neighbors (testnet only)                             |
| `rntn`      | Refused non-tethered neighbors                                          |
| `rtl`       | for transactions randomly removed from the request list                 |
| `lmi`       | Latest milestone index and hash                                         |
| `sn`        | Uses solid milestone's child measurement to publish newly confirmed tx. |
| `tx`        | Newly seen transactions                                                 |
| `vis`       | `tx_hash`, `branch_tx`, `trunk_tx`                                      |
| `lmr`       | list of milestone references                                            |
| `ctx`       | list of finalized transactions                                          |
| `nav`       | newly added validators                                                  |
| `nrv`       | newly removed validators                                                |
| `cvs`       | current validators                                                      |
| `sldf`      | transactions recently solidified by the node                            | 
| `<Address>` | Watching all traffic on a specified address                             |

<!-- [1]: https://javadoc-badge.appspot.com/helixnetwork/helix-1.0.svg?label=javadocs -->

<!-- [2]: https://javadoc-badge.appspot.com/helixnetwork/helix-1.0 -->

[3]: https://img.shields.io/github/v/release/helixnetwork/pendulum?include_prereleases 

<!-- [12]: https://img.shields.io/matrix/helixnetwork:matrix.org.svg?label=matrix -->

<!-- [13]: https://riot.im/app/#/room/#helixnetwork:matrix.org -->

[4]: https://img.shields.io/badge/License-GPLv3-blue.svg

[5]: LICENSE

[6]: https://travis-ci.com/HelixNetwork/pendulum.svg?token=iyim5S8NXU1bnHDx8VMr&branch=master

[7]: https://travis-ci.com/HelixNetwork/pendulum

[8]: https://api.codacy.com/project/badge/Grade/f90eeaff3b1c4e9fb324c74100ed7b3a

[9]: https://www.codacy.com?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=HelixNetwork/pendulum&amp;utm_campaign=Badge_Grade

[10]: https://codecov.io/gh/HelixNetwork/pendulum/branch/dev/graph/badge.svg

[11]: https://codecov.io/gh/HelixNetwork/pendulum

[14]: https://img.shields.io/discord/410771391600656395.svg?label=discord

[15]: https://discord.gg/PjAKR8q


CHANGELOG.md
## 1.0.7
- Fix and refactoring of validation failures in multi-spend value transactions and a corresponding unit-test.

## 1.0.6
- Updated mainnet snapshot and updated the public key as noted in helix-dao-data
- Disable remote API calls: `attachToTangle`, `interruptAttachingToTangle`, `getNeighbors`, `addNeighbors`

## 1.0.5
-  Fixed `getBalance`: `RoundViewModel.get()` returns null on `index`=0, thus NPE was thrown when `references` were not passed and the first round hadn't been completed. In our implementation the snapshot is already constructed based on relative confirmations, thus it suffices for `getBalances` to respond with balance according to `latestSnapshot`.
-  Updated `previousEpochsSpentAddresses` resource files 

## 1.0.4
-  Fixed several zmq publish statements in which an incorrect format was specified.
-  Added `vis`, `lmr` and `ctx` zmq [topics](https://github.com/HelixNetwork/pendulum#messageq) to track basic info for visualisation.
-  Added test for TagViewModel
-  Added default constructor to TagHash to support loading from DB
-  MockitoHint was fixed for new Mockito version
-  Fixed [#215](https://github.com/HelixNetwork/pendulum/issues/215) key rotation: `INCOMPLETE` candidates are re-analyzed, and once solid, the candidate transaction should be processed correctly, and key rotation executed properly.
-  Fixed [#208](https://github.com/HelixNetwork/pendulum/issues/208): When reaching a milestone in the dag-traversal, the branch containing the merkle root is now correctly traversed. Virtual transactions will also solve this issue.
- Changed some log statements to preferred format: Rather than ```metric_label : metric_value``` is better to use metric_label = metric_value because when parsing, we can split on "=", but not ":" since the later is container in the timestamp of the logline
- Removed not-needed trace from state diff tracking

## 1.0.3
-  Integrated `getConfirmationState`: "ConfirmationState" (previously "InclusionState") is computed using `tx.confirmations` and `CONFIRMATION_THRESHOLD`. This is a preliminary modification to enable a more liveness-oriented design, in which a client does not see the states "pending" / "confirmed", but constant updates of the relative confirmations, _until_ a specifiable threshold is reached, at which we consider a transaction confirmed (finalized). Details will be available in the [specifications](https://github.com/HelixNetwork/helix-specs/tree/master/specs/1.0).
-  Added trace logs for balance inconsistency checks #209
-  Added additional logging for upcoming snapshotting changes #210
-  Optimized logging levels and readability
-  Testnet option now uses correct testnet genesis time
-  Fixed dag-traversal in LedgerService in #213

## 1.0.2
-  Added roundIndex to transaction meta data
-  Set round indices of a round's txs in MilestoneTracker
-  Removed obsolete `isPoWDisabled` setting
-  Removed obsolete spammer-class
-  Changed default dir of local snapshots from mainnet to snapshot
-  Only debug log most recent applied round
-  Fixed #184: `NULL_HASH` not used as empty round's snapshot hash anymore
-  Temporarily added parameter `NUM_OF_ACTIVE_VALIDATORS`
-  Temporarily added parameter `CONFIRMATION_THRESHOLD`
-  Set `roundIndex` of all round's txs - not only milestones
-  Update `getConfirmedTransactions`
-  Added `confirmations` to tx meta data. Each time a valid milestone refers to a transaction this counter is incremented.
-  New implementation of `getConfirmationStates`, which now compares relative confirmations to a specifiable confirmation threshold.
-  Add `validator-path` parameter

## 1.0.1
-   Added dev instrumentation
-   Added hardware specs

## 1.0.0
-   Added new implementation of `TransactionTestUtils.buildTransaction()`
-   Added test for DAGHelper
-   Added test for SpentAddressesProvider
-   Update SpentAddressesProviderImpl for multi-testing
-   Uniform random tip selection with config param `TIPSELECTION_ALPHA` = 0
-   Allow `X-HELIX-API-Version` header in CORS
-   Fix for logging-issue #180
-   The config param "port" was called, API_PORT in the APIConfig class, but PORT in the BasePendulumConfig class. To remain consistent, all instances of PORT when API_PORT was meant have been updated.
-   Fixed The logback-save.xml producing some unnecessary messages at the start of a node.  

## 0.6.9
-   Set `GENESIS_TIME` to [mainnet-genesis-time](https://github.com/HelixNetwork/helix-dao-data#genesis-time)
-   Update Undertow from 1.4.6 to 2.0.26
-   Removed obsolete nominee tracking
-   Renamed curator->ValidatorManager
-   Renamed nominee -> Validator
-   Modified the following configs:
    - `--curator` => `--validator-manager`
    - `--update-nominee` => `--update-validator`
    - `--start-nominee` => `--start-validator`
    - `--nominee` => `--validator`
    - `--testnet-no-coo-validation` => `--testnet-no-milestone-sign-validation`
-   Fixed Logback configuration

## 0.6.8
-   Remove Graphstream
-   Snapshots are saved to specifiable dir within of root
-   Updated Global Mainnet Snapshot

## 0.6.7
-   Fixed #142 Negative value for address detected in the running testnet

## 0.6.6
-   Fixed RVM.getRandomMilestone: returns null if round has empty set
-   Fixed for RVM.delete: if round is deleted from DB, it's removed from cache
-   Added test for RoundViewModel

## 0.6.5
-   getNodeInfo now only reflects relevant information
-   Store nominees in local db and fix getNomineesOfRound()
-   Deleted empty HelixTest
-   Fix debug info and javadocs in Tests
-   Fix for debug info, javadoc and variable names in src

## 0.6.4
-   Renaming/Refactoring
-   `round_duration` and `round_pause` was increased for testing

## 0.6.3
-   Added test for Round model
-   Update Round Model
-   Add test for hashes
-   Sha3 getStandardHash() returns zeros for 0-length input array
-   Hashes model was updated to support reading with offset
-   HashPrefix length set to 32

configure-many.py
# -*- coding: utf-8 *-*
# Module: Topology & Neighbors
# This module will handle the static topology of the network
# and generate configuration files for the jars.
# Note that the term, "static", is important, since the overall goal
# should be to acheive a simulation that allows for node joining and
# leaving the network.
import shutil
import os
from subprocess import call
import argparse
import networkx as nx

class SmallWorldTopology():
    def __init__(self, numNodes, averageNeighborsPerNode, probabilityAddingAdditionalNeighbors):
        self.nNeighbors = averageNeighborsPerNode
        self.pNewEdge = probabilityAddingAdditionalNeighbors
        self.overlay = nx.watts_strogatz_graph(n=numNodes, k=self.nNeighbors, p=self.pNewEdge)

    def get_neighbor(self, nodeID):
        if list(self.overlay.neighbors(nodeID)):
            return list(self.overlay.neighbors(nodeID))

    def get_matrix(self):
        self.overlay_matrix = nx.to_numpy_matrix(self.overlay)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='')
    parser.add_argument('-apiPortStart', metavar='Port to start iterating at', type=int, default=14000, help='')
    parser.add_argument('-udpReceiverPortStart', metavar='UDP Receiver port to start iterating at', type=int, default=24000, help='')
    parser.add_argument('-tcpReceiverPortStart', metavar='TCP Receiver port to start iterating at', type=int, default=34000, help='')
    parser.add_argument('-zmqIpcStart', metavar='ZMQ ipc port to start iterating at', type=int, default=1, help='')
    parser.add_argument('-zmqPortStart', metavar='ZMQ port to start iterating at', type=int, default=5550, help='')
    parser.add_argument('-numNodes', metavar='Number of nodes.', type=int, default=10, help='')
    args = parser.parse_args()
    apiPortStart = args.apiPortStart
    udpReceiverPortStart = args.udpReceiverPortStart
    tcpReceiverPortStart = args.tcpReceiverPortStart
    inet = "127.0.0.1"
    tcpInetPort = udpReceiverPortStart
    udpInetPort = udpReceiverPortStart
    zmqIpcStart = args.zmqIpcStart
    zmqPortStart = args.zmqPortStart
    numNodes = args.numNodes
    topology = SmallWorldTopology(numNodes, averageNeighborsPerNode=8, probabilityAddingAdditionalNeighbors=0.1)
    for i in range(numNodes):
        config =  """#NODE {} CONFIG
API_PORT = {}
API_HOST = localhost
REMOTE_LIMIT_API = addNeighbors, getNeighbors, removeNeighbors, attachToTangle, interruptAttachingToTangle
MAX_FIND_TRANSACTIONS = 100_000
MAX_REQUESTS_LIST = 1_000
MAX_GET_BYTES = 10_000
MAX_BODY_LENGTH = 1_000_000
#REMOTE_AUTH =
MS_DELAY = 0
UDP_RECEIVER_PORT = {}
TCP_RECEIVER_PORT = {}
NEIGHBORS
P_REMOVE_REQUEST = 0.01
SEND_LIMIT = -1
MAX_PEERS = 0
DNS_REFRESHER_ENABLED = true
DNS_RESOLUTION_ENABLED = true
HXI_DIR = hxi
DB_PATH = mainnetdb
DB_LOG_PATH = mainnet.log
DB_CACHE_SIZE = 100_000
ROCKS_DB = rocksdb
REVALIDATE = false
RESCAN_DB = false
P_REPLY_RANDOM_TIP = 0.66
P_DROP_TRANSACTION = 0.0
P_SELECT_MILESTONE_CHILD = 0.7
P_SEND_MILESTONE = 0.02
P_PROPAGATE_REQUEST = 0.01
MWM = 1
PACKET_SIZE = 1200
REQ_HASH_SIZE = 32
QUEUE_SIZE = 1_000
P_DROP_CACHE_ENTRY = 0.02
CACHE_SIZE_BYTES = 150_000
ZMQ_THREADS = 1
ZMQ_IPC = ipc://sbx/feeds/{}
ZMQ_ENABLED = false
ZMQ_PORT = {}
GRAPH_ENABLED = false
MAX_DEPTH = 15
ALPHA = 0.001
TIP_SOLIDIFIER_ENABLED = true
POW_THREADS = 0
COORDINATOR_ADDRESS = 6a8413edc634e948e3446806afde11b17e0e188faf80a59a8b1147a0600cc5db
LOCAL_SNAPSHOTS_ENABLED = true
LOCAL_SNAPSHOTS_PRUNING_ENABLED = true
LOCAL_SNAPSHOTS_PRUNING_DELAY = 50000
LOCAL_SNAPSHOTS_INTERVAL_SYNCED = 10
LOCAL_SNAPSHOTS_INTERVAL_UNSYNCED = 1000
LOCAL_SNAPSHOTS_BASE_PATH = mainnet
LOCAL_SNAPSHOTS_DEPTH = 100
SNAPSHOT_FILE = /snapshotMainnet.txt
SNAPSHOT_SIG_FILE = /snapshotMainnet.sig
PREVIOUS_EPOCHS_SPENT_ADDRESSES_TXT = /previousEpochsSpentAddresses.txt
PREVIOUS_EPOCHS_SPENT_ADDRESSES_SIG = /previousEpochsSpentAddresses.sig
GLOBAL_SNAPSHOT_TIME = 1522235533
MILESTONE_START_INDEX = 0
NUM_KEYS_IN_MILESTONE = 10
MAX_ANALYZED_TXS = 20_000
SAVELOG = true
SAVELOG_BASE_PATH = logs/
SAVELOG_XML_FILE = /logback-save.xml
""".format(i, apiPortStart,udpReceiverPortStart,tcpReceiverPortStart,zmqIpcStart, zmqPortStart)
        nodeNeighbors = ""
        for n in topology.get_neighbor(i):
            nodeNeighbors = nodeNeighbors + "tcp://{}:{},".format(inet, tcpInetPort+n)
        nodeNeighbors = nodeNeighbors[:-1]
        config = config.replace("NEIGHBORS", "NEIGHBORS = {}".format(nodeNeighbors))
        apiPortStart +=1
        udpReceiverPortStart +=1
        tcpReceiverPortStart +=1
        zmqIpcStart +=1
        zmqPortStart +=1
        numNodes +=1
        if os.path.isdir(os.path.normpath("./many_configs")):
            pass
        else:
            os.mkdir(os.path.normpath("./many_configs"))
        with open(os.path.normpath("./many_configs/config{}.ini".format(i)), "w") as f:
            f.write(config)
